---
title: "Rapid expressed variant and allelic bias detection for rare variants and rare diseases"
author: ""
date: "5/10/2023"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Contact Infomation

## Installing

To install the private package:

    # install.packages("devtools")
    library(devtools)
    
    # Bioconductor needs to be activated when automatically installing dependencies.
    options(repos = BiocManager::repositories())
    devtools::install_github("trichelab/bamSliceR")
    
    library(bamSliceR)

## If you will be analyzing any data that is controlled access, you will need to download a GDC token and store it in a file in your home directory. For more instructions see [GenomicDataCommons Token](https://www.bioconductor.org/packages/devel/bioc/vignettes/GenomicDataCommons/inst/doc/overview.html#authentication) and [Data portal authentication tokens](https://docs.gdc.cancer.gov/Data_Portal/Users_Guide/Cart/#gdc-authentication-tokens)

```{bash}
echo "my_downloaded_auth_code" > ~/.gdc_token
```

## Inspect the available projects on GDC portal
#### Check ids for all the projects on GDC portal
```{r message=FALSE, warning=FALSE}
library(bamSliceR)
availableProjectId()
```
#### Check the available Experimental Strategies give a project id.
```{r message=FALSE}
availableExpStrategy("TARGET-AML")
```
#### Check the available Workflows for each Experimental Strategies of a project.
```{r message=FALSE}
availableWorkFlow(projectId = "TARGET-AML", es = "RNA-Seq")
```
#### Get the information of BAM files
After knowing the keyword of project, experimental strategy and workflow, we can
collect information of BAMs we are interested.  
```{r message=FALSE}
file_meta = getGDCBAMs(projectId = "TARGET-AML", es = "RNA-Seq", workflow = "STAR 2-Pass Genome")
head(file_meta)
```
#### Example on how to make the character() vector describing chromosomal regions.
```{r message=FALSE}
target_genes = readRDS("./data/target_genes.rds")
target_genes
```
Get granges for exons of all genes
```{r message=FALSE}
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(Homo.sapiens)
library(stringr)

txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
#all the GDC files are against hg38,
TxDb(Homo.sapiens) <- TxDb.Hsapiens.UCSC.hg38.knownGene

exs <- exonsBy(Homo.sapiens, "gene", columns="SYMBOL")
names(exs) <- mapIds(Homo.sapiens, names(exs), "SYMBOL", "GENEID")
exs <- exs[which(!is.na(names(exs)))]
```
Get the exons of target genes and make the format for bamSliceR input
```{r message=FALSE}
target_genes_exs = target_genes[which(names(target_genes) %in% names(exs)) ]
target_genes_exons = exs[names(target_genes_exs) %>% unique()]
target_ranges = reduce(unlist(target_genes_exons) )
target_ranges = subset(target_ranges, seqnames %in% paste0 ("chr", c(1:22, "X", "Y") ))
target_ranges_chars = paste0(as.character(seqnames(target_ranges)), ":", start(ranges(target_ranges)), "-", end(ranges(target_ranges)) )
head(target_ranges_chars)
```

## Downloading the sliced BAMs 
```{r message=FALSE}
#Download 3225 RNA-Seq sliced BAMs files with reads within regions defined by "target_ranges_chars" from TARGET-AML.
#downloadSlicedBAMs(file_df = file_meta, regions = target_ranges_chars, dir = "./inst/extdata/")
```

## Tally the reads of sliced BAM files
We first also need to specify the regions as a GRanges object. 
```{r message=FALSE}
head(target_ranges)
```
Then we need make the character() vector including all the names of downloaded BAM files.
I normally would do:
```{bash, eval=FALSE}
cd DIR_BAM_FILES
ls | grep bam$ > bamfiles
```
in the directory of the downloaded BAM files. And then scan the 'bamfiles' in R.
```{r message=FALSE}
#bamfiles = scan("bamfiles", "character")
```
Last thing we need is to specify the directory of gmapGenome object created before.
```{r message=FALSE}
#gmapGenome_dir = "/varidata/research/projects/triche/TARGET/GMKF/oncohistone/BAMs/hg38/"
```
We can then start to tally the reads of BAMs files:
```{r message=FALSE}
#tallyReads(bamfiles = bamfiles, gmapGenome_dir = gmapGenome_dir, grs = target_ranges,
#           BPPARAM = MulticoreParam(workers = 10 , stop.on.error = TRUE), parallelOnRanges = TRUE,
#           parallelOnRangesBPPARAM = MulticoreParam(workers = 10) )
```
## Tricks on running bamSliceR on HPC with multiple nodes
TallyVariants() originally from package 'VariantTools' allows parallelizing computing on both BAM files and GRanges using the same parameter, BPPARAM, for bplapply(). The setting is not efficient on certain circumstances and sometimes would unnecessarily eat up all the memory. For example, if we want to tally reads on thousands of BAMs but only few gene regions, ideally we want to put more workers on parallelizing computing on BAM files but less workers on granges regions.
To overcome the issue, I modified TallyVariants() and wrapper up it to tallyReads(). We can now specify if we want to parallelize computing on granges regions using parameter "parallelOnRanges", and provide the 'parallelOnRangesBPPARAM' specific for parallelize computing on granges regions. 

Example1: tally reads on "chr17:7665307:7704652" of 679 BAMs files
If we set 10 workers to parallelize compute on both BAM files and granges regions, it would take ~12 mins to complete. 
```{r message=FALSE, eval=FALSE}
library(bamSliceR)
library(VariantAnnotation)
library(BiocParallel)
library(gmapR)
library(VariantTools)
setwd("/varidata/research/projects/triche/Peter/bamSliceR/TP53/RNA_ALL_BAMs")
bamfiles = scan("bamfiles", "character")
gmapGenome_dir = "/varidata/research/projects/triche/TARGET/GMKF/oncohistone/BAMs/hg38/"

GRanges( seqnames = Rle (c("chr17")) ,  IRanges(start=7665307, end=7704652), strand = Rle(strand(c("*")) ) ) -> TP53_gr
tallyReads(bamfiles = bamfiles, gmapGenome_dir = gmapGenome_dir, grs = TP53_gr,
           BPPARAM = MulticoreParam(workers = 10 , stop.on.error = TRUE), parallelOnRanges = TRUE,
           parallelOnRangesBPPARAM = MulticoreParam(workers = 10) ) -> TARGET_ALL_RNA_TP53

saveRDS(TARGET_ALL_RNA_TP53, "../TARGET_ALL_RNA_TP53.rds")
```
If we set 80 workers to parallelize compute on BAM files and set 'parallelOnRanges = FALSE', it would take ~2 mins to complete. 
```{r message=FALSE, eval=FALSE}
library(bamSliceR)
library(VariantAnnotation)
library(BiocParallel)
library(gmapR)
library(VariantTools)
setwd("/varidata/research/projects/triche/Peter/bamSliceR/TP53/RNA_ALL_BAMs")
bamfiles = scan("bamfiles", "character")
gmapGenome_dir = "/varidata/research/projects/triche/TARGET/GMKF/oncohistone/BAMs/hg38/"

GRanges( seqnames = Rle (c("chr17")) ,  IRanges(start=7665307, end=7704652), strand = Rle(strand(c("*")) ) ) -> TP53_gr
tallyReads(bamfiles = bamfiles, gmapGenome_dir = gmapGenome_dir, grs = TP53_gr,
           BPPARAM = MulticoreParam(workers = 80 , stop.on.error = TRUE), parallelOnRanges = FALSE,
           parallelOnRangesBPPARAM = MulticoreParam(workers = 10) ) -> TARGET_ALL_RNA_TP53

saveRDS(TARGET_ALL_RNA_TP53, "../TARGET_ALL_RNA_TP53_2.rds")
```

Example2: 100 BAM files (sliced on 500+ genes) in Lauren's folder "BAM_slices_T20_09" and 50 gene regions:
```{r message=FALSE, eval=FALSE}
library(bamSliceR)
library(VariantAnnotation)
library(BiocParallel)
library(gmapR)
library(VariantTools)
setwd("/varidata/research/projects/triche/TARGET/GMKF/germline_validations/BAM_slices_T20_09")
bamfiles = scan("/varidata/research/projects/triche/Peter/bamSliceR/TARGET_AML_germline/BAM_slices_T20_09/bamfiles", "character")
gmapGenome_dir = "/varidata/research/projects/triche/TARGET/GMKF/oncohistone/BAMs/hg38/"
target_ranges_gr = readRDS("/varidata/research/projects/triche/Peter/bamSliceR/TARGET_AML_germline/target_ranges_gr.rds")
seqlevelsStyle(target_ranges_gr) <- "UCSC"
keepSeqlevels(target_ranges_gr, paste0("chr", c(1:22,"X") ) ) -> target_ranges_gr

c( "TARGET-20-PASBHI-09A-01R.RNA.GRCh38.sliced.bam",
 "TARGET-20-PASVYA-09A-01R.RNA.GRCh38.sliced.bam",
 "TARGET-20-PASVYL-09A-01R.RNA.GRCh38.sliced.bam",
 "TARGET-20-PASWLN-09A-01R.RNA.GRCh38.sliced.bam") -> noIndexBamfiles
bamfiles[ -which(bamfiles %in% noIndexBamfiles)] -> indexed_bamfiles

Sys.time() -> t1
tallyReads(bamfiles = indexed_bamfiles[1:100] , gmapGenome_dir = gmapGenome_dir, grs = target_ranges_gr[1:50],
           BPPARAM = MulticoreParam(workers = 10 , stop.on.error = TRUE), parallelOnRanges = TRUE,
                parallelOnRangesBPPARAM = MulticoreParam(workers = 10 ) ) -> TARGET_AML_RNA_10
Sys.time() -> t2
t2 - t1

#Time difference of 52.10761 mins
```

#### Template on submitting bamSliceR jobs on HPC
We can use sbatch to submit bamSliceR job in Rcode to new HPC.
The Rcode for downloading BAMs would be like this:
```{r message=FALSE, eval = FALSE}
#bamSliceR_Download.r
library(bamSliceR)
library(GenomicDataCommons)
library(httr)

BAMs_FOLDER_DIR = "/varidata/research/projects/triche/Peter/bamSliceR/WT1/TARGET_AML_RNA"
WT1 = c("chr11:32379149-32468665")
TARGET_AML_RNA_BAMs = getGDCBAMs("TARGET-AML", "RNA-Seq", "STAR 2-Pass Genome" )

downloadSlicedBAMs(file_df = TARGET_AML_RNA_BAMs, regions = WT1, dir = BAMs_FOLDER_DIR)
```
The Rcode for tally BAMs would be like this:
```{r message=FALSE, eval=FALSE}
#bamSliceR_TallyReads.r
library(bamSliceR)
library(GenomicDataCommons)
library(VariantAnnotation)
library(BiocParallel)

BAMs_FOLDER_DIR = "/varidata/research/projects/triche/Peter/bamSliceR/WT1/TARGET_AML_RNA"
gmapGenome_dir = "/varidata/research/projects/triche/TARGET/GMKF/oncohistone/BAMs/hg38/"
setwd(BAMs_FOLDER_DIR)

WT1_gr = GRanges( seqnames = Rle (c("chr11")) ,  IRanges(start=32379149, end=32468665), strand = Rle(strand(c("*")) ) )
TARGET_AML_RNA_BAMs = getGDCBAMs("TARGET-AML", "RNA-Seq", "STAR 2-Pass Genome" )

badbamfiles = scan("bad_bams.fofn", "character")
bamfiles = scan("bamfiles", "character")
if (length(badbamfiles) == 0)
{
    valid_bamfiles = bamfiles
} else
{
    valid_bamfiles = bamfiles[ -which( bamfiles %in% badbamfiles)]
}
    
tallyReads(bamfiles = valid_bamfiles, gmapGenome_dir = gmapGenome_dir, grs = WT1_gr,
           BPPARAM = MulticoreParam(workers = 80 , stop.on.error = TRUE), parallelOnRanges = FALSE,
           parallelOnRangesBPPARAM = MulticoreParam(workers = 10) ) -> TARGET_AML_RNA_WT1
saveRDS(TARGET_AML_RNA_WT1, "TARGET_AML_RNA_WT1.rds")
TARGET_AML_RNA_WT1_combined = stackSamples(VRangesList(TARGET_AML_RNA_WT1))

TARGET_AML_RNA_WT1_combined = annotateWithBAMinfo(tallied_reads = TARGET_AML_RNA_WT1_combined, file_meta = TARGET_AML_RNA_BAMs)
saveRDS(TARGET_AML_RNA_WT1_combined, "TARGET_AML_RNA_WT1_combined.rds")
```
The sbatch file would be like this (bamSliceR.sh):
```{bash, eval=FALSE}
#!/bin/bash

#SBATCH --export=NONE
#SBATCH -J TallyReads2
#SBATCH -o TallyReads2.o
#SBATCH -e TallyReads2.e
#SBATCH --ntasks 1
#SBATCH --time 3:00:00
#SBATCH --mem=800G
BAMs_FOLDER_DIR="/varidata/research/projects/triche/Peter/bamSliceR/WT1/TARGET_AML_RNA/"
start_time=$(date +"%T")

R CMD BATCH $BAMs_FOLDER_DIR/bamSliceR_Download.r

end_time=$(date +"%T")
echo "Downloading BAMs:"
echo "Start time: $start_time"
echo "End time: $end_time"

start_time=$(date +"%T")

cd $BAMs_FOLDER_DIR
samtools quickcheck -v *.bam > bad_bams.fofn   && echo 'all ok' || echo 'some files failed check, see bad_bams.fofn'
for i in $(ls | grep bam$); do samtools index $i; done
ls | grep bam$ > bamfiles

end_time=$(date +"%T")
echo "Index BAMs:"
echo "Start time: $start_time"
echo "End time: $end_time"


start_time=$(date +"%T")

R CMD BATCH $BAMs_FOLDER_DIR/bamSliceR_TallyReads.r

end_time=$(date +"%T")
echo "Tally BAMs:"
echo "Start time: $start_time"
echo "End time: $end_time"
```
Submit the jobs to hpc:
```{bash, eval=FALSE}
sbatch -p bigmem -n 128 bamSliceR.sh
```
