---
title: "Slicing and Tallied BAMs for Interested Genes and Variant Analysis"
author:
    name: Peter Huang
    affiliation: Van Andel InstituteMichigan State University
date: "13 July 2024"
output: 
  rmarkdown::html_vignette:
    toc: true
    fig_caption: true
bibliography: document.bib
csl: bioinformatics.csl
link-citations: true
vignette: >
  %\VignetteIndexEntry{BamSliceRWorkFlow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Abstract
In this vignette, we will walk through the bamSliceR toolset to swiftly extract 
coordinate- or range-based aligned DNA and/or RNA sequence reads from the Genomic 
Data Commons [GDC](https://portal.gdc.cancer.gov/). We will then demonstrate the 
downstream functionality to tally and annotate variants from BAM files, whether 
they are stored in the cloud or locally. Also, we want to highlight the new 
utilities we implemented in the `bamSliceR` to facilitate transcript-aware 
variants annotation using transcriptome BAM that generated by aligning reads 
from RNA-seq to reference transcripts sequences (GENCODE v36).

We will use data from TARGET [@Bolouri2017-kp], BEAT-AML [@Tyner2018-ur] and 
Leucegene [@Lavallee2016-wx].

# Introduction
We developed `bamSliceR` to address two practical challenges: resource-sparing 
identification of candidate subjects, and variant detection from aligned 
sequence reads across thousands of controlled-access subjects. The GDC 
[@Wilson2017-oc] BAM Slicing API is a practical and well-documented REST API for 
this purpose. [@Morgan2017-zj]. Additionally, the `GenomicDataCommons` 
Bioconductor package implements the service in R interface, allowing for more 
efficient querying of data from the GDC. However, we find relatively little 
published work that uses the GDC API, even those evaluating specific candidate 
variants. Instead, researchers often rely on variant calls from previous studies, 
assuming a single best method for variant detection fits all experimental 
designs—a notion contradicted by various benchmarks. Alternatively, retrieving
and reanalyze raw sequence data, which is highly inefficient.

# Set-up

```{r setup0, message=FALSE, echo = FALSE}
options(digits=3)
options(width=90)
```

```{r downloads, message=FALSE}
```

```{r setup, eval=TRUE, message=FALSE}
library(bamSliceR)
```

# `bamSliceR` pipeline

## Download Bam Slices
The [data model for the GDC is complex](https://gdc.cancer.gov/developers/gdc-data-model). If you want query more 
other types of data, it may worth to overview the details. Here, we focus on
identifying the metadata of BAM files, followed by downloading the sliced BAM.

![Graph Representation of the GDC Data Model](GDC_data_model.png){width=550px}

#### Identify the BAM file of interest
We simplified the querying process in a single function to gather BAM files 
information from the project of interest. 

```{r baminfo}
file_meta = getGDCBAMs(projectId = "TARGET-AML", 
                       es = "RNA-Seq", 
                       workflow = "STAR 2-Pass Genome")
nrow(file_meta)
head(file_meta)
```

Three pieces of information are needed to locate the BAM files on GDC portal, 
which can be inspected by `availableProjectId()`, `availableExpStrategy()` and 
`availableWorkFlow()`, respectively.

1) The project ID. 
2) Experiment Strategy (ex. “RAN-Seq”, “WGS”, etc. ) 
3) Alignment workflow.

```{r inspection,message=FALSE, warning=FALSE}
availableProjectId() %>% head(n = 10)
availableExpStrategy("TARGET-AML")
availableWorkFlow(projectId = "TARGET-AML", es = "RNA-Seq")
```
#### Identify the Genomic Ranges of interest
BAM slicing API from GDC portal accept genomic ranges specifying as vector of 
character() e.g., c("chr", "chr1:10000"). Here we provide a function to get the 
required input format given the gene names.

```{r genes,message=FALSE}
target_genes_data = system.file("data", "gene_names.rds", package = "bamSliceR")
target_genes = readRDS(target_genes_data)
target_genes
```

Get either GRanges or vector of character() for exons of the genes.

```{r ranges,message=FALSE}
#Get GRanges for exons of all genes above
target_ranges_gr = getGenesCoordinates(target_genes, ret = "GRanges")
head(target_ranges_gr)

#Get the vector of character() instead.
target_ranges_chars = getGenesCoordinates(target_genes, ret ="DF", 
                                          extendEnds = 100)
head(target_ranges_chars)
```

#### Download sliced-BAM
```{r download, message=FALSE, eval = FALSE}
downloadSlicedBAMs(file_df = file_meta, 
                   regions = target_ranges_chars, 
                   dir = "BAM_FILES")
```

## Tally Variants from sliced BAMs
`bamSliceR` integrated functionality in Bioconductor package `gmapR` to tally 
coverage and counts of variant alleles followed by estimation of Variant Allele 
Fraction (VAF) of each variants.

```{r tallyreads, eval=TRUE}
```

#### Identify the Genomic Ranges of interest
We first also need to specify the regions as a `GRanges` object, which can be
a subset of genomic ranges used to sliced BAM files.

```{r message=FALSE}
head(target_ranges_gr)
```

Then we need make a text file to include all the names of downloaded 
BAM files.

![bamfiles](bamfiles.png){width=400px height=200px}

```{bash, eval=FALSE}
cd DIR_BAM_FILES
ls | grep bam$ > bamfiles
```

In the directory of the downloaded BAM files, we then `scan` the 'bamfiles' into R.

```{r message=FALSE, eval = FALSE}
bamfiles_txt = system.file("extdata", "bamfiles", package = "bamSliceR")
bamfiles = scan("bamfiles", "character")
```

Last thing we need is to specify the directory of gmapGenome object created before.
(see [here](https://github.com/trichelab/bamSliceR/blob/main/vignettes/How_to_create_gmapGenome.r) about how to create gmapGenome object.)

```{r message=FALSE, eval = FALSE}
gmapGenome_dir = "/path/to/your/gmapGenome"
#samtools faidx genome.fa
#samtools faidx genome.fa chr1:100-200
#
```

We can then start to tally the reads of BAMs files:

```{r message=FALSE, eval = FALSE}
tallied_reads = tallyReads(bamfiles = bamfiles, gmapGenome_dir = gmapGenome_dir, grs = target_ranges,
                           BPPARAM = MulticoreParam(workers = 4 , stop.on.error = TRUE), parallelOnRanges = TRUE,
                           parallelOnRangesBPPARAM = MulticoreParam(workers = 4) )
```

For more information about how to efficiently run tallyReads() on HPC on large cohorts,
please see the instruction in [here](https://github.com/trichelab/bamSliceR/blob/main/vignettes/How_to_TallyingRead_On_Parallel.md).


# Data pre-processing 

# Mutation analysis 

## Mutation discovery


## Mutation calling


# Differential Transcript Experssion analysis


# Differential Transcript Usage analysis


## Visualizations for DTU


```{r softwareinfo}
sessionInfo()
```

# References
