---
title: "Slicing and Tallied BAMs for Interested Genes and Variant Analysis"
author:
    name: Peter Huang
    affiliation: Van Andel InstituteMichigan State University
date: "13 July 2024"
output: 
  rmarkdown::html_vignette:
    toc: true
    fig_caption: true
bibliography: document.bib
csl: bioinformatics.csl
link-citations: true
vignette: >
  %\VignetteIndexEntry{BamSliceRWorkFlow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Abstract
In this vignette, we will walk through the bamSliceR toolset to swiftly extract 
coordinate- or range-based aligned DNA and/or RNA sequence reads from the Genomic 
Data Commons [GDC](https://portal.gdc.cancer.gov/). We will then demonstrate the 
downstream functionality to tally and annotate variants from BAM files, whether 
they are stored in the cloud or locally. Also, we want to highlight the new 
utilities we implemented in the `bamSliceR` to facilitate transcript-aware 
variants annotation using transcriptome BAM that generated by aligning reads 
from RNA-seq to reference transcripts sequences (GENCODE v36).

We will use data from TARGET [@Bolouri2017-kp], BEAT-AML [@Tyner2018-ur] and 
Leucegene [@Lavallee2016-wx].

# Introduction
We developed `bamSliceR` to address two practical challenges: resource-sparing 
identification of candidate subjects, and variant detection from aligned 
sequence reads across thousands of controlled-access subjects. The GDC 
[@Wilson2017-oc] BAM Slicing API is a practical and well-documented REST API for 
this purpose. [@Morgan2017-zj]. Additionally, the `GenomicDataCommons` 
Bioconductor package implements the service in R interface, allowing for more 
efficient querying of data from the GDC. However, we find relatively little 
published work that uses the GDC API, even those evaluating specific candidate 
variants. Instead, researchers often rely on variant calls from previous studies, 
assuming a single best method for variant detection fits all experimental 
designs—a notion contradicted by various benchmarks. Alternatively, retrieving
and reanalyze raw sequence data, which is highly inefficient.

# Set-up

```{r setup0, message=FALSE, echo = FALSE}
options(digits=3)
options(width=90)
```

```{r downloads, message=FALSE}
```

```{r setup, eval=TRUE, message=FALSE}
library(bamSliceR)
```

# `bamSliceR` pipeline

# Working with Genomic BAM files

## Download Bam Slices
The [data model for the GDC is complex](https://gdc.cancer.gov/developers/gdc-data-model). If you want query more 
other types of data, it may worth to overview the details. Here, we focus on
identifying the metadata of BAM files, followed by downloading the sliced BAM.

![Graph Representation of the GDC Data Model](GDC_data_model.png){width=550px}

#### Identify the BAM file of interest
We simplified the querying process in a single function to gather BAM files 
information from the project of interest. 

```{r baminfo}
file_meta = getGDCBAMs(projectId = "TARGET-AML", 
                       es = "RNA-Seq", 
                       workflow = "STAR 2-Pass Genome")
nrow(file_meta)
head(file_meta)
```

Three pieces of information are needed to locate the BAM files on GDC portal, 
which can be inspected by `availableProjectId()`, `availableExpStrategy()` and 
`availableWorkFlow()`, respectively.

1) The project ID. 
2) Experiment Strategy (ex. “RAN-Seq”, “WGS”, etc. ) 
3) Alignment workflow.

```{r inspection,message=FALSE, warning=FALSE}
availableProjectId() %>% head(n = 10)
availableExpStrategy("TARGET-AML")
availableWorkFlow(projectId = "TARGET-AML", es = "RNA-Seq")
```
#### Identify the Genomic Ranges of interest
BAM slicing API from GDC portal accept genomic ranges specifying as vector of 
character() e.g., c("chr", "chr1:10000"). Here we provide a function to get the 
required input format given the gene names.

```{r genes,message=FALSE}
target_genes_data = system.file("data", "gene_names.rds", package = "bamSliceR")
target_genes = readRDS(target_genes_data)
target_genes
```

Get either GRanges or vector of character() for exons of the genes.

```{r ranges,message=FALSE}
#Get GRanges for exons of all genes above
target_ranges_gr = getGenesCoordinates(target_genes, ret = "GRanges")
head(target_ranges_gr)

#Get the vector of character() instead.
target_ranges_chars = getGenesCoordinates(target_genes, ret ="DF", 
                                          extendEnds = 100)
head(target_ranges_chars)
```

#### Download sliced-BAM
```{r download, message=FALSE, eval = FALSE}
downloadSlicedBAMs(file_df = file_meta, 
                   regions = target_ranges_chars, 
                   dir = "BAM_FILES")
```

## Tally Variants from sliced BAMs
`bamSliceR` integrated functionality in Bioconductor package `gmapR` to tally 
coverage and counts of variant alleles followed by estimation of Variant Allele 
Fraction (VAF) of each variants.

```{r tallyreads, eval=TRUE}
```

#### Identify the Genomic Ranges of interest
We first also need to specify the regions as a `GRanges` object, which can be
a subset of genomic ranges used to sliced BAM files.

```{r message=FALSE}
head(target_ranges_gr)
```

#### Specify list of BAM files 
Then we need make a text file to include all the names of downloaded 
BAM files.

![bamfiles](bamfiles.png){width=400px height=200px}

```{bash, eval=FALSE}
cd DIR_BAM_FILES
ls | grep bam$ > bamfiles
```

In the directory of the downloaded BAM files, we then `scan` the 'bamfiles' into R.

```{r message=FALSE, eval = FALSE}
bamfiles_txt = system.file("extdata", "bamfiles", package = "bamSliceR")
bamfiles = scan(bamfiles_txt, "character")
```

#### Preparing `gmapGenome` Object
Last thing we need is to create a `gmapGenome` object, then specify the directory 
of it.

```{r message=FALSE, eval = FALSE}
#GmapGenome objects can be created from FASTA files or BSgenome objects. Below is 
#an example of how to create human hg38 gmapGenome from BSgenome.

library(gmapR)
library(BSgenome.Hsapiens.UCSC.hg38)

gmapGenomePath <- file.path("PATH/to/SAVE/gmapGenome", "hg38")
gmapGenomeDirectory <- GmapGenomeDirectory(gmapGenomePath, create = TRUE)

gmapGenome <- GmapGenome(genome=Hsapiens, 
                         directory=gmapGenomeDirectory,
                         name="hg38", create=TRUE) 
```

#### Tallying Reads
We can then start to tally the reads of BAM files.
If you want to run `tallyReads` on machine with multiple nodes, there are 
parameters to distribute the nodes according to number of BAM files and 
number of genomic ranges:

1) `parallelOnRanges`: if `TRUE`, will process the Genomic Ranges in parallel.
2) `parallelOnRangesBPPARAM`: configuration for parallel tallyreads on genomic ranges.
3) `BPPARAM`: configuration for parallel tallyreads on BAM files.

If we want to tally reads on thousands of BAM but only few gene regions, 
ideally we want to put more workers on computing on BAM files but 
less workers on granges regions.

```{r message=FALSE, eval = FALSE}
tallied_reads = tallyReads(bamfiles = bamfiles, gmapGenome_dir = gmapGenome_dir, 
                           grs = target_ranges,
                           BPPARAM = MulticoreParam(workers = 4 , stop.on.error = TRUE),
                           parallelOnRanges = TRUE,
                           parallelOnRangesBPPARAM = MulticoreParam(workers = 4) )
```

We also provide a bash template to run `tallyReads()` on HPC on large cohorts.

## Annotation of Variants tallied from Genomic BAM files

### Annotation of Variants ###
A VRanges object will be generated from tallying reads from BAM files, contains all the putative variants. sampleNames() can be used to see the name of BAM files which variants detected from. Here, We present an example on how to annotate variants with predicted consequence using [VariantAnnotation](https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html) and Ensembl Variant Effect Predictor ([VEP](https://useast.ensembl.org/info/docs/tools/vep/index.html)). 

We started from a VRanges output from tallyReads().
```{r message=FALSE,  warning=FALSE, echo='hide'}
library(bamSliceR)

tallied_reads = system.file("data", "tallied_reads_example.rds", 
                            package = "bamSliceR")

tallied_reads_vr = readRDS(tallied_reads)

# VRanges-specific methods such as altDepth(), refDepth(), totalDepth() would not
# availiable after conversion to GRanges. So save those info now.
tallied_reads_vr = saveVRinfo(tallied_reads_vr)

# Match back the metadata of BAM files to the VRanges
file_meta = getGDCBAMs("TARGET-AML", "RNA-Seq", "STAR 2-Pass Genome")
tallied_reads_vrinfo_baminfo = annotateWithBAMinfo(tallied_reads_vr, file_meta)

# Only keep variants with variant allele frequency greater than 5%.
tallied_reads_vrinfo_baminfo = subset(tallied_reads_vrinfo_baminfo, VAF > 0.05)

```

#### VariantAnnotation ####
Consequence of variants now can be predicted using [VariantAnnotation](https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html):

```{r message=FALSE,  warning=FALSE, echo='hide'}

getVariantAnnotation(tallied_reads_vrinfo_baminfo) -> tallied_reads_vrinfo_baminfo_gr

```
#### Ensembl Variant Effect Predictor (VEP) ####
Consequence of variants also can be predicted using Variant Effect Predictor ([VEP](https://useast.ensembl.org/info/docs/tools/vep/index.html)).
The code below shows how to generate VCF file as input for VEP:
```{r message=FALSE,  warning=FALSE, echo='hide'}

gr2vrforVEP(tallied_reads_vrinfo_baminfo, file = "~/INPUT_VCF_FILE.vcf", 
            writeToVcf = TRUE) -> vr

head(vr)
```

Details about how to run Variant Effect Predictor can be found in ([ensemblVEP](https://bioconductor.org/packages/release/bioc/html/ensemblVEP.html)) or ([VEP](https://useast.ensembl.org/info/docs/tools/vep/index.html)).

The VCF file with variant effect predicted can be annotated back to patients' variants.
```{r message=FALSE,  warning=FALSE, echo='hide', eval=FALSE}
#Example vcf file with variant effect predicted.
# Output VCF file from VEP
vep_file = system.file("extdata", "TARGET_AML_VRforVEP_vep.vcf", package = "bamSliceR")

#Extract the predicted consequences of variants from VEP
csqFromVEP = getCSQfromVEP(vep_file)

#Annotated the variants with VEP predicted consequences
tallied_reads_vrinfo_baminfo_annotated_gr = getVEPAnnotation(tallied_reads_vrinfo_baminfo_gr, 
                                                             csqFromVEP)

```

### Visualization ###
To facilitate downstream analysis of variants from patients, bamSliceR provides customized 
plotVAF() function based on the maftools::plotVaF() from [maftools](https://www.bioconductor.org/packages/release/bioc/vignettes/maftools/inst/doc/maftools.html)
to help user to investigate the Variant Allele Frequency of the selected variants.

Also, bamSliceR provides utility that allow conversion of GRanges to MAF format file, which is compatible with maftools.
Oncoplots, survival analysis, and mutual exclusive test, etc, can then be implemented easily.

#### Distribution of VAF ####

By default, the plotVAF would plot the distribution of VAF in top genes.
```{r message=FALSE,  warning=FALSE, echo='hide' }
library(bamSliceR)

TARGET_AML_RNA_annotated_file = system.file("data", "TARGET_AML_RNA_annotated.gr.rds", 
                            package = "bamSliceR")

tallied_reads_vrinfo_baminfo_annotated_gr = readRDS(TARGET_AML_RNA_annotated_file)
plotVAF(tallied_reads_vrinfo_baminfo_annotated_gr, title = "Default: VAF of Top genes")

```

By specifying single gene and multiple coordinates against the gene products, the plotVAF()
would plot distribution of VAF in selected loci of the gene.
```{r message=FALSE,  warning=FALSE, echo='hide' }
plotVAF(tallied_reads_vrinfo_baminfo_annotated_gr, genes = "IDH2", 
        bySingleLocus = c(140, 172), title = "VAF of IDH2 at R140 and R172")

```

By specifying multiple genes and single coordinate against the genes' products, the plotVAF()
would plot distribution of VAF in selected loci of the gene.
```{r message=FALSE,  warning=FALSE, echo='hide' }

histone_genes <- c(
  HIST1H3A = "H3C1",
  HIST1H3B = "H3C2",
  HIST1H3C = "H3C3",
  HIST1H3D = "H3C4",
  HIST1H3E = "H3C6",
  HIST1H3F = "H3C7",
  HIST1H3G = "H3C8",
  HIST1H3H = "H3C10",
  HIST1H3I = "H3C11",
  HIST1H3J = "H3C12", H3F3A = "H3-3A"
)

plotVAF (tallied_reads_vrinfo_baminfo_annotated_gr, 
         genes = histone_genes, 
         bySingleLocus = c(28), title = "VAF of Histone 3 genes at K27")

```

Plotting the VAFf distribution of two hotspot H3K27M & H3K27I, which were previously
documented have deleterious effect in pediatric brain tumor.
```{r message=FALSE,  warning=FALSE, echo='hide' }

plotVAF (tallied_reads_vrinfo_baminfo_annotated_gr, 
         genes = histone_genes, 
         groupByAAchanges = c("K28M", "K28I"), 
         title = "VAF of Histone 3 genes at K27M/I")

```

### Covert to MAF file format ###

Example of how to convert GRanges to MAF format file.
```{r message=FALSE,  warning=FALSE, echo='hide' }
TARGET_AML_RNA_annotated_filtered_file = system.file("data", 
                                                   "TARGET_AML_RNA_annotated_filtered.rds", 
                                                   package = "bamSliceR")
TARGET_AML_RNA_annotated_filtered.gr = readRDS(TARGET_AML_RNA_annotated_filtered_file)

```

#### Keep sample from sinlg-run and one timepoint ####
Patients in TARGET-AML usually have multiple sequencing runs from multiple timepoints.
The downstream statistical analysis from maftools would NOT take account the multi-run and 
multi-timepoints senario for each patients. In this case, when converting to MAF format,
we recommend user to specify the sample type and only keep single sequencing run of the patient.

All the options of sample types in GDC data base can be view as below:
```{r message=FALSE,  warning=FALSE, echo='hide' }

bamSliceR:::GDC_SAMPLE_TYPE

```

We now collect the variants of samples derived from primary disease.
```{r message=FALSE,  warning=FALSE, echo='hide' }

TARGET_AML_RNA_annotated_primary_gr = keepSampleType(TARGET_AML_RNA_annotated_filtered.gr, 
                                                     sample_type = "primary", 
                                                     keepUniquePatient = TRUE)
table(TARGET_AML_RNA_annotated_primary_gr$sample_type)

```

#### GRanges to MAF format ####
Now we can convert the GRanges to MAF format.
```{r message=FALSE,  warning=FALSE, echo='hide' }

grToMAF(TARGET_AML_RNA_annotated_primary_gr, 
        file = "~/TARGET_AML_RNA_annotated_primary_maf.tsv") -> TARGET_AML_RNA_annotated_primary_maf

head(TARGET_AML_RNA_annotated_primary_maf)

```

#### Prepare Clinical Data ####
Clinical information is optional but recommended in maftools, because many downstream analysis
is depended on.
```{r message=FALSE,  warning=FALSE, echo='hide' }

TARGET_AML_clinical_file = system.file("extdata", "TARGET_AML_clinical_maf.tsv", package = "bamSliceR")

```

#### Maintain Sample Size for read.maf() ####
One last step before using maftools::read.maf() to create a maf object, We need 
to add some dummy variants
to the MAF file, so that the read.maf() would not filtered the patients without 
any variants. For example, the clinical data includes 500 patients, but for the 
genes we selected, there are 200 patients don't have any variants or mutations 
in those selected genes. In this case, read.maf() would shrink the sample size 
from 500 to 300, because those 200 patients without any variants would not present 
in the input maf file. (more information can be found [here](https://github.com/PoisonAlien/maftools/issues/159).).

To maintain the sample size when using read.maf(), we provide generateDummyVariants() to
add dummy variants to maf file. These dummy variants will not affect the results from 
downstream analysis performed by maftools, if user correctly specify the 'vc_nonSyn'
with valid type of variants.

Example:
```{r message=FALSE,  warning=FALSE, echo='hide' }

# Collect the ID of all patients
TARGET_AML_clinical = read.delim(TARGET_AML_clinical_file)
TARGET_all_patient_IDs = TARGET_AML_clinical$Tumor_Sample_Barcode

# Generate dummy variants for those patients without any mutations.
TARGET_AML_RNA_annotated_primary_all_patients_maf = generateDummyVariants(patients_ID = TARGET_all_patient_IDs, 
                                                                          maf.df = TARGET_AML_RNA_annotated_primary_maf,
                                                                          file = "~/TARGET_maf_dummyAdded.tsv")

head(TARGET_AML_RNA_annotated_primary_all_patients_maf)
```

#### Create maf object  ####
Now, we can create MAF object using maftools::read.maf :
```{r message=FALSE,  warning=FALSE, echo='hide', message=FALSE}
library(maftools)
# Valid or nonSyn type of variants, exlucde the "dummy" entity.
target.maf.nonSyn = unique(TARGET_AML_RNA_annotated_primary_maf$Variant_Classification)

# Path to maf file
TARGET_maf_file = system.file("extdata", "TARGET_maf_dummyAdded.tsv",
                              package = "bamSliceR")
# Path to clinical data
TARGET_clinical_file = system.file("extdata", "TARGET_AML_clinical_maf.tsv",
                                  package = "bamSliceR")

TARGET_maf = read.maf(maf = TARGET_maf_file, 
                      clinicalData = TARGET_clinical_file, 
                      vc_nonSyn = target.maf.nonSyn)

TARGET_maf
```

#### Downstream analysis from maftools  ####

Example of maftools::somaticInteractions().
```{r message=FALSE,  warning=FALSE, echo='hide', message=FALSE}

me = somaticInteractions(maf = TARGET_maf, top = 50, pvalue = c(0.05, 0.1))

```
Example of maftools::oncoplot().
```{r message=FALSE,  warning=FALSE, echo='hide' }

oncoplot(maf = TARGET_maf, top = 10)

```

Example of maftools::mafSurvival().
```{r message=FALSE,  warning=FALSE, echo='hide' }

mafSurvival(maf = TARGET_maf, genes = c("DNMT3A"), groupNames = c("DNMT3A R882H", "WT"), 
            time = 'days_to_last_followup', Status = 'Overall_Survival_Status', isTCGA = TRUE)

```

Example of maftools::mafSurvival() on combined cohorts of TARGET-AML and BEAT-AML.
```{r message=FALSE,  warning=FALSE, echo='hide' }


TARGET_BEAT_AML_maf_file = system.file("extdata", "TARGET_BEAT_AML_maf_dummyAdded.tsv",
                              package = "bamSliceR")
TARGET_BEAT_AML_clinical_file = system.file("extdata", "TARGET_BEAT_AML_clinical.tsv",
                                 package = "bamSliceR")

all_maf = read.maf(maf = TARGET_BEAT_AML_maf_file, 
        clinicalData = TARGET_BEAT_AML_clinical_file, vc_nonSyn = target.maf.nonSyn)

mafSurvival(maf = all_maf, genes = c("DNMT3A"), groupNames = c("DNMT3A R882H", "WT"), 
            time = 'days_to_last_followup', Status = 'Overall_Survival_Status', isTCGA = TRUE)

```

# Mutation analysis 

## Mutation discovery


## Mutation calling


# Differential Transcript Experssion analysis


# Differential Transcript Usage analysis


## Visualizations for DTU


```{r softwareinfo}
sessionInfo()
```

# References
