BamSlicing
================
5/10/2023

## Installation instructions

## Inspect the available projects on GDC portal

#### Check ids for all the projects on GDC portal

```{r message=FALSE, warning=FALSE, echo='hide'}
library(BamSlicing)
availableProjectId()
```

#### Check the available Experimental Strategies give a project id.

```{r message=FALSE}
availableExpStrategy("TARGET-AML")
```

#### Check the available Workflows for each Experimental Strategies of a project.

```{r message=FALSE}
availableWorkFlow(projectId = "TARGET-AML", es = "RNA-Seq")
```

## Case Study

### Pediatric Acute Myeloid Leukemia (AML) onco-histone mutations
To investigate the genetic landscape of lysine-to-Methionine mutation at 27th animo acid of Histone 3 genes (H3K27M) in AML across ages, we examined 11 histone 3 genes as well as genes encoding epigenetic factors that are frequently mutated in AML patients (*IDH1/2*, *DNMT3A*, *RUNX1*, *ASXL1/2* and *TET1/2*). 

#### Query metadata of BAM files of Project on GDC
We can use availableProjectId(), availableExpStrategy() and availableWorkFlow() to find the keywords in order to get the information of BAM files on GDC portal. Here, we showed the example using TARGET-AML cohort with 2,281 subjects including 3,225 RNA-seq BAM files.

```{r message=FALSE}
file_meta = getGDCBAMs(projectId = "TARGET-AML", es = "RNA-Seq", workflow = "STAR 2-Pass Genome")
head(file_meta)
```

#### Make the character() vector of genomic regions for BAM slicing.
BAM slicing API from GDC portal accept genomic ranges specifying as vector of character() e.g., c("chr", "chr1:10000"). Here we provide a function to get the required input format given the gene names.

```{r message=FALSE}
library(VariantAnnotation)
genes = readRDS("./data/gene_names.rds")
genes

#Get GRanges for exons of all genes above
genes_gr = getGRangesGivenGeneNames(genes, exons = TRUE, reduce = TRUE)
head(genes_gr)

#Get the vector of character() instead.
genes_chars = getGRangesGivenGeneNames(genes, exons = TRUE, as.character = TRUE, reduce = TRUE)
head(genes_chars)
```

#### Downloading the sliced BAMs 
The GDC offers both “controlled-access” and “open” data. As of this writing, all BAM files on GDC are "controlled-access". Details on how to request and setup a authentication token, please see here: https://bioconductor.org/packages/release/bioc/vignettes/GenomicDataCommons/inst/doc/overview.html#authentication

Once token are setup correctly, sliced BAM files can be downloaded using downloadSlicedBAMs().

```{r message=FALSE}
#Download 3225 RNA-Seq sliced BAMs files with reads within regions defined by "target_ranges_chars" from TARGET-AML.
#downloadSlicedBAMs(file_df = file_meta, regions = target_ranges_chars, dir = "./inst/extdata/")
```

#### Tally the reads of sliced BAM files

Details on how to download sliced BAM files followed by tallying the reads, see the README.

#### Annotation of Variants ####
A VRanges object will be generated from tallying reads from BAM files, contains all the putative variants. sampleNames() can be used to see the name of BAM files which variants detected from. Here, We present an example on how to annotate variants with predicted consequence using [VariantAnnotation](https://bioconductor.org/packages/release/bioc/html/VariantAnnotation.html) and Ensembl Variant Effect Predictor ([VEP](https://useast.ensembl.org/info/docs/tools/vep/index.html)). 

